### PR TITLE
Fix IO throttling during copying whole directories

### LINKED ISSUE #52183: Bandwidth configration not work for moving data from local disk to s3 by ttl
clickhouse version:22.3.1

I tried Mergetree with S3 disks. Data stored in local disks and then moved to s3 by ttl. 
I limited 
`                        <max_remote_read_network_bandwidth_for_server>10485760</max_remote_read_network_bandwidth_for_server>
                     <max_remote_write_network_bandwidth_for_server>10485760</max_remote_write_network_bandwidth_for_server>
                       <max_replicated_fetches_network_bandwidth_for_server>10485760</max_replicated_fetches_network_bandwidth_for_server>
                     <max_replicated_sends_network_bandwidth_for_server>10485760</max_replicated_sends_network_bandwidth_for_server>
<max_network_bandwidth>10485760</max_network_bandwidth>
<max_network_bandwidth_for_all_users>10485760</max_network_bandwidth_for_all_users>`

No people use this cluster except clickhouse ttl moving parts.But the bandwidth more than 10Mb/s
image

_Originally posted by @coulsonm in https://github.com/ClickHouse/ClickHouse/issues/52079#issuecomment-1637487505_

### LINKED ISSUE #52753: Can we limit bandwidth for s3 disk?
I don't understand why max_..._network_bandwidth don't work for me.

```sql
SELECT * FROM system.disks
┌─name────┬─path──────────────────────────────┬───────────free_space─┬──────────total_space─┬─────unreserved_space─┬─keep_free_space─┬─type──┬─is_encrypted─┬─is_read_only─┬─is_write_once─┬─is_remote─┬─is_broken─┬─cache_path─┐
│ default │ /var/lib/clickhouse/              │         105359450112 │         105491914752 │         105359450112 │               0 │ local │            0 │            0 │             0 │         0 │         0 │            │
│ s3disk  │ /var/lib/clickhouse/disks/s3disk/ │ 18446744073709551615 │ 18446744073709551615 │ 18446744073709551615 │               0 │ s3    │            0 │            0 │             0 │         1 │         0 │            │
└─────────┴───────────────────────────────────┴──────────────────────┴──────────────────────┴──────────────────────┴─────────────────┴───────┴──────────────┴──────────────┴───────────────┴───────────┴───────────┴────────────┘
```

```sql
CREATE TABLE test_network_bandwidth_s3
(
    `A` Int64,
    `S` String,
    `D` Date
)
ENGINE = MergeTree
PARTITION BY D
ORDER BY A
SETTINGS storage_policy = 's3tiered', old_parts_lifetime=1;

insert into test_network_bandwidth_s3 select number, number, '2023-01-01' from numbers(1e9);

optimize table test_network_bandwidth_s3 final;

select disk_name, partition, sum(rows), formatReadableSize(sum(bytes_on_disk)) size, count() part_count
from system.parts where table= 'test_network_bandwidth_s3' and active
group by disk_name, partition;

┌─disk_name─┬─partition──┬──sum(rows)─┬─size─────┬─part_count─┐
│ s3disk    │ 2023-01-01 │ 1000000000 │ 7.64 GiB │          1 │
└───────────┴────────────┴────────────┴──────────┴────────────┘


alter table test_network_bandwidth_s3 move partition '2023-01-01' to volume 's3disk';
0 rows in set. Elapsed: 100.791 sec.


┌─disk_name─┬─partition──┬──sum(rows)─┬─size─────┬─part_count─┐
│ s3disk    │ 2023-01-01 │ 1000000000 │ 7.64 GiB │          1 │
└───────────┴────────────┴────────────┴──────────┴────────────┘

alter table test_network_bandwidth_s3 move partition '2023-01-01' to volume 'default';
0 rows in set. Elapsed: 123.183 sec.
```
data is moved to S3:  Elapsed: 100.791 sec.
data is moved from S3: Elapsed: 123.183 sec.

Let's try to limit traffic

```sql
SELECT name, value FROM system.settings WHERE name LIKE 'max_rem%';
┌─name──────────────────────────────────────────┬─value──┐
│ max_remote_read_network_bandwidth             │ 0      │
│ max_remote_write_network_bandwidth            │ 0      │
│ max_remote_read_network_bandwidth_for_server  │ 100000 │
│ max_remote_write_network_bandwidth_for_server │ 100000 │
└───────────────────────────────────────────────┴────────┘

SELECT uptime();
┌─uptime()─┐
│       31 │
└──────────┘


alter table test_network_bandwidth_s3 move partition '2023-01-01' to volume 's3disk';
0 rows in set. Elapsed: 114.517 sec.

alter table test_network_bandwidth_s3 move partition '2023-01-01' to volume 'default';
0 rows in set. Elapsed: 100.739 sec.
```

data is moved to S3:  Elapsed: 114.517 sec.
data is moved from S3: Elapsed: 100.739 sec.
```

Let's limit more

```sql
┌─name──────────────────────────────────────────┬─value─┐
│ max_remote_read_network_bandwidth             │ 50000 │
│ max_remote_write_network_bandwidth            │ 50000 │
│ max_remote_read_network_bandwidth_for_server  │ 50000 │
│ max_remote_write_network_bandwidth_for_server │ 50000 │
└───────────────────────────────────────────────┴───────┘

SELECT uptime()
┌─uptime()─┐
│       23 │
└──────────┘

alter table test_network_bandwidth_s3 move partition '2023-01-01' to volume 's3disk';
0 rows in set. Elapsed: 102.329 sec.


alter table test_network_bandwidth_s3 move partition '2023-01-01' to volume 'default';
0 rows in set. Elapsed: 109.543 sec.
```

data is moved to S3:  Elapsed: 102.329 sec.
data is moved from S3: Elapsed: 109.543 sec.

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix IO throttling during copying whole directories (MOVE PARTITION TO VOLUME, fetching/cloning parts, mutations, ...)

**P.S. help with tests is highly appreciated** (_for now I've tested only locally `MOVE PARTITION TO VOLUME`_)

Fixes: #52753 (cc @den-crane)
Fixes: #52183 (cc @coulsonm)

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit d349988838ae09e9c27966e8671a5a72892e2405 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53338/d349988838ae09e9c27966e8671a5a72892e2405/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by coulsonm** [comment]
image

FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/actions-runner/_work/_temp/tests_bugfix_check/integration/integration_commit_status.tsv'

**Comment by coulsonm** [comment]
@azat

**Comment by azat** [comment]
This is likely internal CI issue. And this PR does not have a test yet.

**Comment by coulsonm** [comment]
May I ask when will it be repaired?

**Comment by azat** [comment]
This should not be required for merging this PR, it just needs some attention. @serxa maybe you can take a look?

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp [hunk 1]
```diff
@@ -769,8 +769,11 @@ void DiskObjectStorageTransaction::createFile(const std::string & path)
         }));
 }
 
-void DiskObjectStorageTransaction::copyFile(const std::string & from_file_path, const std::string & to_file_path)
+void DiskObjectStorageTransaction::copyFile(const std::string & from_file_path, const std::string & to_file_path, const WriteSettings & settings)
 {
+    /// NOTE: For native copy we can ignore throttling, so no need to use WriteSettings
+    (void)settings;
```
Comment (serxa) [inline src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:775]:
```suggestion
    UNUSED(settings);
```

**Review by serxa — APPROVED** [review]
Looks good. As for tests, at least a simple check for the duration of explicit ALTER TABLE MOVE PARTITION would be good enough.

**Comment by azat** [comment]
> As for tests, at least a simple check for the duration of explicit ALTER TABLE MOVE PARTITION would be good enough.

Yeah, I was just lazy and thought maybe someone will contribute the test in the mean time, but this did not happen, so I've added the test.

**Pull request merged**